src/libostree/ostree-repo-commit.c

```diff
@@ -408,18 +453,176 @@ add_size_index_to_metadata (OstreeRepo        *self,
+      g_hash_table_remove_all (self->object_sizes);
     }
 
   return g_variant_ref_sink (g_variant_builder_end (builder));
 }
 
+static gboolean
+read_sizes_entry (GVariant                     *entry,
+                  OstreeContentSizeCacheEntry  *sizes,
+                  char                         *csum,
+                  GError                      **error)
+{
+  gsize entry_size = g_variant_get_size (entry);
+  g_return_val_if_fail (entry_size >= 34, FALSE);
+
+  const guchar *buffer = g_variant_get_data (entry);
+  if (buffer == NULL)
+    return glnx_throw (error, "Could not read ostree.sizes metadata entry");
+
+  ostree_checksum_inplace_from_bytes (buffer, csum);
+  buffer += 32;
+  entry_size -= 32;
+
+  gsize bytes_read = 0;
+  guint64 archived = 0;
+  if (!_ostree_read_varuint64 (buffer, entry_size, &archived, &bytes_read))
+    return glnx_throw (error, "Unexpected EOF reading ostree.sizes varint");
+  sizes->archived = archived;
+  buffer += bytes_read;
+  entry_size -= bytes_read;
+
+  guint64 unpacked = 0;
+  if (!_ostree_read_varuint64 (buffer, entry_size, &unpacked, &bytes_read))
+    return glnx_throw (error, "Unexpected EOF reading ostree.sizes varint");
+  sizes->unpacked = unpacked;
+  buffer += bytes_read;
+  entry_size -= bytes_read;
+
+  /* On newer commits, an additional byte is used for the object type. */
+  if (entry_size > 0)
+    {
+      OstreeObjectType objtype = *buffer;
+      if (objtype < OSTREE_OBJECT_TYPE_FILE || objtype > OSTREE_OBJECT_TYPE_LAST)
+        return glnx_throw (error, "Unexpected ostree.sizes object type %u",
+                           objtype);
+      sizes->objtype = objtype;
+    }
+  else
+    {
+      /* Assume the object is a file. */
+      sizes->objtype = OSTREE_OBJECT_TYPE_FILE;
+    }
+
+  return TRUE;
+}
+
+/**
+ * ostree_repo_get_commit_sizes
+ * @self: Self
+ * @rev: Commit checksum
+ * @new_archived: (out) (optional): number of archived bytes for the commit
+ *                missing from the repository, or %NULL
+ * @new_unpacked: (out) (optional): number of unpacked bytes for the commit
+ *                missing from the repository, or %NULL
+ * @new_objects: (out) (optional): number of objects for the commit missing
+ *               from the repository, or %NULL
+ * @archived: (out) (optional): number of archived bytes for the commit in the
+ *            repository, or %NULL
+ * @unpacked: (out) (optional): number of unpacked bytes for the commit in the
+ *            repository, or %NULL
+ * @objects: (out) (optional): number of objects for the commit in the
+ *           repository, or %NULL
+ * @cancellable: a #GCancellable
+ * @error: a #GError
+ *
+ * Reads the size data for the commit stored in the %ostree.sizes key in
+ * the commit metadata. If this data is not available, %FALSE is
+ * returned and @error is set to %G_IO_ERROR_NOT_FOUND.
+ *
+ * Returns: %TRUE on success, %FALSE on failure
+ *
+ * Since: 2019.5
+ */
```

Comment:
This is the API as we have it in Endless, but I feel like it's not quite right. What this function actually returns is a summary of the `ostree.sizes` data, so something like `ostree_repo_get_commit_size_summary()` might make more sense. Also, the 6 out parameters seems excessive. Should this be a struct?

This API only returns a summary, but there's nothing that gets you the entries. Parsing the entries manually is tricky because of the variable length integers, and `_ostree_read_varuint64()` is not exposed publicly. I think an `ostree_repo_read_commit_sizes()` returning an array of `OstreeContentSizeCacheEntry` (possibly renamed) would also be nice so that the program could like through all the entries instead of just getting the summary.

Reply:
I'd name this something more like `ostree_repo_get_missing_object_sizes()` or so.

Having a struct instead of the out parameters makes sense to me.

I wonder whether it'd be more ergonomic though for this to be part of `OstreeAsyncProgress` - during a pull if e.g. `OSTREE_REPO_PULL_FLAG_CALCULATE_MISSING_SIZES` is passed, we provide all of this data as part of the progress?

All callers who want this would have to basically do a pull to get the commit, then call this API, then determine if they wanted to resume the pull, right?

Reply:
> I'd name this something more like `ostree_repo_get_missing_object_sizes()` or so.

I can't think of any really great names, although the `missing` is a bit of a misnomer because it returns both object size information as well as the number of objects in both total and local counts. That's why I suggested `get_commit_size_summary`, but this is almost certainly going to be used get the size of the objects you don't have. Maybe `get_commit_object_summary` since it's not just sizes. Feel strongly one way or the other?

> Having a struct instead of the out parameters makes sense to me.

:+1:

> I wonder whether it'd be more ergonomic though for this to be part of `OstreeAsyncProgress` - during a pull if e.g. `OSTREE_REPO_PULL_FLAG_CALCULATE_MISSING_SIZES` is passed, we provide all of this data as part of the progress?
> 
> All callers who want this would have to basically do a pull to get the commit, then call this API, then determine if they wanted to resume the pull, right?

That's interesting. The way it's used in our updater now is that we always do a metadata only pull and then use the API to get statistics about the full pull. That data is then exported over DBus to callers that want to present it like the About page in control center and possibly in GNOME software.

@pwithnall any thoughts about what this API should look like?  What's in the PR is what eos-updater is using now.

Reply:
How about this. Instead of figuring out the best way to use this summarizing API right now, I just add an API that parses the entries and returns them in an array? Then the caller can do what they want since the summarizing part is simple once you can read the entries but application specific. I think that the ability to parse the checksums, variable length integers and object type are needed regardless.

Over in eos-updater we can inline the calculation it needs (it ignores the number of object out params now anyways) after calling the new entries parsing API to get an array of entries.

Reply:
Sorry I’m late to the party. From my point of view, either is fine. Adding an API to OSTree to just return the metadata seems like a reasonable compromise if people can’t agree on semantics for things otherwise. From the eos-updater point of view, the important things are:
 * Not relying on custom downstream libostree API
 * Not ending up with a different, forked implementation of functionality which exists upstream (getting commit sizes)

So going with a metadata-only API would be fine as long as we didn’t then deviate on how to store and calculate the summaries from it.

Reply:
Currently eos-updater doesn't store anything about the summaries. It just passes the size totals out as DBus properties. And the size totals are simple to calculate - just loop over each of the entries, add the sizes for that entry to the "full" totals, check if the local repo has the object in the entry, add the sizes to the "needed" totals if the object doesn't exist locally.

To me the important part is that we get the changes to the commit metadata upstream:

* Include all objects in the sizes array
* Include the object type in each size entry so you can reliably determine if you have the object corresponding to the entry or not

Once the metadata has this information and it can be parsed easily, what to do with it is really application specific.

Reply:
> That's why I suggested get_commit_size_summary

I am having a mental hurdle using `summary` for that since it's not related to the summary file.  `get_commit_size_info()`?

Reply:
I dropped this API in the most recent changes. Instead, there's a lower level API that just parses the metadata and returns an array representing it. The `show` builtin has an example of summing the metadata as this API was doing before. It's pretty trivial and what I plan to do in eos-updater (except that it doesn't care about all the sums, so parts of it can be left out).